storagenode/piecestore/endpoint.go

```diff
@@ -282,6 +282,10 @@ func (endpoint *Endpoint) Upload(stream pb.Piecestore_UploadServer) (err error)
 			if err := endpoint.VerifyPieceHash(ctx, limit, message.Done, expectedHash); err != nil {
 				return err // TODO: report grpc status internal server error
 			}
+			if message.Done.PieceSize != pieceWriter.Size() {
+				return ErrProtocol.New("Size of finished piece does not match size declared by uplink! %d != %d",
+					message.Done.GetPieceSize(), pieceWriter.Size())
+			}
 
```

Comment:
Let's also check that the timestamp is reasonable.

Reply:
I would agree, but I'm not sure what reasonable necessarily means here. I think the storagenode will already have to keep track of the exact sent timestamp in pieceinfo (the piece file timestamp on the filesystem probably doesn't have enough resolution and may not be entirely reliable under all possible operating conditions), so I don't much care that the uplink sends any timestamp in particular. If it wants to send 0 or something in 1989, the storagenode can be fine with that if we want it to be. The size is more important- we _can_ use the size of the piece file to keep track of what goes in the PieceHash, as long as we make sure it was right to begin with.